### PR TITLE
check for if-some-or-ok-else-none-or-err

### PR DESCRIPTION
**Pull request by J-ZhengLi**
fixes: #8492 

---

changelog: make [`option_if_let_else`] to check for match expression with both Option and Result; **TODO: Change lint name? Add new lint with similar functionality?**

### CODE REVIEW

File: clippy_lints/src/option_if_let_else.rs [hunk 1]
```diff
@@ -99,43 +119,39 @@ fn format_option_in_sugg(cx: &LateContext<'_>, cond_expr: &Expr<'_>, as_ref: boo
...
-        if !expr.span.from_expansion(); // Don't lint macros, because it behaves weirdly
-        if !in_constant(cx, expr.hir_id);
-        if let Some(higher::IfLet { let_pat, let_expr, if_then, if_else: Some(if_else) })
-            = higher::IfLet::hir(cx, expr);
-        if !is_else_clause(cx.tcx, expr);
-        if !is_result_ok(cx, let_expr); // Don't lint on Result::ok because a different lint does it already
-        if let PatKind::TupleStruct(struct_qpath, [inner_pat], _) = &let_pat.kind;
-        if is_lang_ctor(cx, struct_qpath, OptionSome);
-        if let PatKind::Binding(bind_annotation, _, id, None) = &inner_pat.kind;
+        if !is_result_ok(cx, cond_expr); // Don't lint on Result::ok because a different lint does it already
```
Comment (J-ZhengLi) [inline clippy_lints/src/option_if_let_else.rs:134]: This was with the original code, I wonder what lint it was refering to? cuz I think removing it does nothing new but I was too afraid to remove it. @flip1995

**Comment by Serial-ATA** [comment]
Should the lint be renamed if it's going to cover `match` now as well?

**Comment by Serial-ATA** [comment]
I think this should only check for arms that span a single line.

I don't think going from

```rust
    let _ = match res {
        Err(_) => 1,
        Ok(mut a) => {
            a += 1;
            println!("{a}");
            a -= 1;
            let b = 5;
            a += b;
            a
        }
    };
```

to

```rust
    let _ = res.map_or(1, |mut a| {
            a += 1;
            println!("{a}");
            a -= 1;
            let b = 5;
            a += b;
            a
        });
```

makes much of a difference.

**Comment by flip1995** [comment]
Thanks! Sorry for taking so long. I rebased it and did a minor cleanup. About the lint name: Yes it will need to be changed. But since this lint is currently in `nursery` there's no rush to decide this now. I added a FIXME comment for it.

@bors r+

**Pull request merged**